crates/shared/src/request_id.rs

```diff
@@ -0,0 +1,84 @@
+//! any behaviour in the process.
+//! Instead we use task local storage that is globally visible but only
+//! individual to each task. That way we can populate the storage with the
+//! identifier once and not care about dragging it through the code base.
+//! And when we issue requests to another process we can simply fetch the
+//! current identifier specific to our task and send that along with the
+//! request.
+tokio::task_local! {
+    pub static REQUEST_ID: std::cell::RefCell<String>;
+}
+
+/// Tries to read the `request_id` from this task's storage.
+/// Returns `None` if task local storage was not initialized or is empty.
+pub fn get_task_local_storage() -> Option<String> {
+    let mut id = None;
+    let _ = REQUEST_ID.try_with(|cell| {
+        id = Some(cell.borrow().clone());
+    });
+    id
+}
+
+/// Tries to write the `request_id` to the task local storage.
+/// Fails if this task has no task local storage.
+pub fn set_task_local_storage(request_id: String) {
+    let _ = REQUEST_ID.try_with(|storage| {
+        *storage.borrow_mut() = request_id;
+    });
+}
+
+/// Takes a `tower::Service` and embeds it in a `make_service` function that
+/// spawns one of these services per incoming request.
+/// But crucially before spawning that service task local storage will be
+/// initialized with some request id.
+/// Either that gets taken from the requests `X-REQUEST-ID` header of if that's
+/// missing a globally unique request number will be generated.
+#[macro_export]
+macro_rules! make_service_with_task_local_storage {
+    ($service:expr) => {{
+        {
+            let internal_request_id = Arc::new(std::sync::atomic::AtomicUsize::new(0));
+            hyper::service::make_service_fn(move |_| {
+                let warp_svc = $service.clone();
+                let internal_request_id = internal_request_id.clone();
+                async move {
+                    let svc =
+                        hyper::service::service_fn(move |req: hyper::Request<hyper::Body>| {
+                            let mut warp_svc = warp_svc.clone();
+                            let id = if let Some(header) = req.headers().get("X-Request-ID") {
+                                String::from_utf8_lossy(header.as_bytes()).to_string()
+                            } else {
+                                format!(
+                                    "{}",
+                                    internal_request_id
+                                        .fetch_add(1, std::sync::atomic::Ordering::SeqCst)
+                                )
+                            };
+                            let span = tracing::info_span!("request", id);
+                            let handle_request = shared::request_id::REQUEST_ID.scope(
+                                std::cell::RefCell::new(id),
+                                async move {
+                                    // Not sure why but we have to have this async block to avoid
```

Comment:
A bit scary. Wouldn't we expect the compiler to optimise this away?

Reply:
Yeah, I would have expected this code to basically be equivalent to simply passing `hyper::service::Service::call()` (without `.await`).
My guess is that either moving data into the future or explicitly having an `.await` point disables some optimization but I'm not sure.
I guess it would be worth it to research if this behavior is guaranteed to stay like this or if this code could break if the compiler becomes smarter in the future.